File: matlab/+IK/IK_2_intersecting.m
function [Q, is_LS_vec] = IK_2_intersecting(R_06, p_0T, kin)
% Axes 5 and 6 intersect
Q = [];
is_LS_vec = [];

p_06 = p_0T - kin.P(:,1) - R_06*kin.P(:,7);

[q4_vec, soln_num_vec] = search_1D( ...
    @(q4)(alignment_err_given_q4(q4, p_06, R_06, kin)), ...
    -pi, pi, 200, false);

for i_q4 = 1:length(q4_vec)
    [e, Q_partial] = alignment_err_given_q4(q4_vec(i_q4), p_06, R_06, kin);
    q_partial = Q_partial(:,soln_num_vec(:,i_q4));

    R_04 = rot(kin.H(:,1),q_partial(1)) * rot(kin.H(:,2),q_partial(2)) ...
         * rot(kin.H(:,3),q_partial(3)) * rot(kin.H(:,4),q_partial(4));
    [q5, q5_is_LS] = subproblem.sp_1(kin.H(:,6), R_04'*R_06*kin.H(:,6),  kin.H(:,5));
    [q6, q6_is_LS] = subproblem.sp_1(kin.H(:,5), R_06'*R_04*kin.H(:,5), -kin.H(:,6));
    q_i = [q_partial; q5; q6];
    Q = [Q q_i];
    is_LS_vec = [is_LS_vec [q5_is_LS; q6_is_LS; e(soln_num_vec(:,i_q4))] ];
end

end

function [e_vec, Q_partial] = alignment_err_given_q4(q4, p_16, R_06, kin)
    e_vec = NaN([1 4]);
    Q_partial = NaN([4 4]);
    % find up to 4 solutions of (q1, q2, q3) using Subproblem 5
    p_35_3 = kin.P(:,4) + rot(kin.H(:,4), q4)*kin.P(:,5);
    
    [t1, t2, t3] = subproblem.sp_5( ...
                   -kin.P(:,2), p_16, kin.P(:,3), p_35_3, ...
                   -kin.H(:,1), kin.H(:,2), kin.H(:,3));

    for i_q123 = 1:length(t1)
        q1 = t1(i_q123);
        q2 = t2(i_q123);
        q3 = t3(i_q123);

        R_04 = rot(kin.H(:,1),q1) * rot(kin.H(:,2),q2) ...
             * rot(kin.H(:,3),q3) * rot(kin.H(:,4),q4);

        e_i = kin.H(:,5)' * R_04' * R_06* kin.H(:,6) - kin.H(:,5)'*kin.H(:,6);
        e_vec(i_q123) = e_i;
        Q_partial(:,i_q123) = [q1; q2; q3; q4];
    end
end

File: cpp/IK/IK_2_intersecting.cpp
#include <iostream>
#include "IK_2_intersecting.h"
#include "../subproblems/sp.h"
#include <eigen3/Eigen/src/Core/Matrix.h>

Eigen::Matrix4d q_partial_given_q4(double q4, const Kinematics &kin,
                                   const Eigen::Vector3d &p_16) {
    Eigen::Matrix4d q_partial;
    q_partial.setConstant(NAN);

    Eigen::Vector3d p_35_3 = kin.P.col(3) + rot(kin.H.col(3), q4) * kin.P.col(4);

    std::vector<double> t1;
    std::vector<double> t2;
    std::vector<double> t3;
    IKS::sp5_run(-kin.P.col(1), p_16, kin.P.col(2), p_35_3, -kin.H.col(0),
                 kin.H.col(1), kin.H.col(2), t1, t2, t3);

    for (unsigned i = 0; i < t1.size(); ++i) {
        Eigen::Vector4d col;
        col << t1[i], t2[i], t3[i], q4;
        q_partial.col(i) = col;
    }

    return q_partial;
}

Solution IK_2_intersecting(const Eigen::Matrix<double, 3, 3> &R_06,
                           const Eigen::Vector3d &p_0T, const Kinematics &kin) {
    Solution soln;

    Eigen::Vector3d p_16 = p_0T - kin.P.col(0) - R_06 * kin.P.col(6);

    auto alignment_error_given_q4 = [&R_06, &kin, &p_16](double q4) {
        Eigen::Vector4d error;
        error.fill(INFINITY);

        Eigen::Vector3d p_35_3 =
            kin.P.col(3) + rot(kin.H.col(3), q4) * kin.P.col(4);

        std::vector<double> t1;
        std::vector<double> t2;
        std::vector<double> t3;
        IKS::sp5_run(-kin.P.col(1), p_16, kin.P.col(2), p_35_3, -kin.H.col(0),
                     kin.H.col(1), kin.H.col(2), t1, t2, t3);

        for (unsigned i = 0; i < t1.size(); ++i) {
            Eigen::Matrix3d r_04 = rot(kin.H.col(0), t1[0]) *
                rot(kin.H.col(1), t2[0]) *
                rot(kin.H.col(2), t3[0]) * rot(kin.H.col(3), q4);

            error[i] =
                (kin.H.col(4).transpose() * r_04.transpose() * R_06 * kin.H.col(5) -
                 kin.H.col(4).transpose() * kin.H.col(5))(0);
        }

        return error;
    };

    std::vector<std::pair<double, unsigned>> q4_vec =
        search_1d<4>(alignment_error_given_q4, -M_PI, M_PI, 200);

    for (std::pair<double, unsigned> zero : q4_vec) {
        double q4 = zero.first;
        unsigned i = zero.second;

        Eigen::Vector4d q_partial = q_partial_given_q4(q4, kin, p_16).col(i);

        Eigen::Matrix3d r_04 =
            rot(kin.H.col(0), q_partial[0]) * rot(kin.H.col(1), q_partial[1]) *
            rot(kin.H.col(2), q_partial[2]) * rot(kin.H.col(3), q_partial[3]);

        double q5;
        bool q5_is_ls =
            IKS::sp1_run(kin.H.col(5), (r_04.transpose() * R_06 * kin.H.col(5)),
                         kin.H.col(4), q5);

        double q6;
        bool q6_is_ls =
            IKS::sp1_run(kin.H.col(4), (R_06.transpose() * r_04 * kin.H.col(4)),
                         -kin.H.col(5), q6);

        Eigen::Matrix<double, 6, 1> q;
        q << q_partial[0], q_partial[1], q_partial[2], q_partial[3], q5, q6;
        soln.q.push_back(q);
        soln.is_ls.push_back(q5_is_ls || q6_is_ls);
    }

    return soln;
}
